pkg/controller/elasticsearch/filesettings/file_settings.go

```diff
@@ -0,0 +1,164 @@
+	if p.Spec.Elasticsearch.ClusterSettings != nil {
+		state.ClusterSettings = p.Spec.Elasticsearch.ClusterSettings
+	}
+	if p.Spec.Elasticsearch.SnapshotLifecyclePolicies != nil {
+		state.SLM = p.Spec.Elasticsearch.SnapshotLifecyclePolicies
+	}
+	/*if p.Spec.Elasticsearch.SecurityRoleMappings != nil {
+		state.RoleMappings = p.Spec.Elasticsearch.SecurityRoleMappings
+	}
+	if p.Spec.Elasticsearch.AutoscalingPolicies != nil {
+		state.Autoscaling = p.Spec.Elasticsearch.AutoscalingPolicies
+	}
+	if p.Spec.Elasticsearch.IndexLifecyclePolicies != nil {
+		state.IndexLifecyclePolicies = p.Spec.Elasticsearch.IndexLifecyclePolicies
+	}
+	if p.Spec.Elasticsearch.IngestPipelines != nil {
+		state.IngestPipelines = p.Spec.Elasticsearch.IngestPipelines
+	}
+	if p.Spec.Elasticsearch.IndexTemplates != nil {
+		if state.IndexTemplates == nil {
+			state.IndexTemplates = &IndexTemplates{}
+		}
+		if p.Spec.Elasticsearch.IndexTemplates.ComposableIndexTemplates != nil {
+			state.IndexTemplates.ComposableIndexTemplates = p.Spec.Elasticsearch.IndexTemplates.ComposableIndexTemplates
+		}
+		if p.Spec.Elasticsearch.IndexTemplates.ComponentTemplates != nil {
+			state.IndexTemplates.ComponentTemplates = p.Spec.Elasticsearch.IndexTemplates.ComponentTemplates
+		}
+	}*/
+	s.State = state
+	return nil
+}
+
+// mutateSnapshotRepositorySettings ensures that a snapshot repository can be used across multiple ES clusters.
+// The Elasticsearch cluster name is injected in the repository settings depending on the type of the repository:
+// - "azure", "gcs", "s3": the `base_path` property is added is set to the cluster name
+// - "fs": the cluster name is appended to the `location` property
+// - "hdfs": the cluster name is appended to the `path` property
+// - "url": nothing is done, the repository is read-only
+// - "source": nothing is done, the repository is an indirection to another repository
+func mutateSnapshotRepositorySettings(snapshotRepository map[string]interface{}, esNs string, esName string) (map[string]interface{}, error) {
+	untypedSettings := snapshotRepository["settings"]
+	if untypedSettings == nil {
+		untypedSettings = map[string]interface{}{}
+	}
+
+	suffix := fmt.Sprintf("%s-%s", esNs, esName)
+	settings, ok := untypedSettings.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("invalid type for snapshot repository settings")
+	}
+	switch snapshotRepository["type"] {
+	case "azure", "gcs", "s3":
+		settings["base_path"] = fmt.Sprintf("snapshots/%s", suffix)
+	case "fs":
+		location, ok := settings["location"].(string)
+		if !ok {
+			return nil, errors.New("invalid type for snapshot repository location")
+		}
+		settings["location"] = filepath.Join(location, suffix)
+	case "hdfs":
+		path, ok := settings["path"].(string)
+		if !ok {
+			return nil, errors.New("invalid type for snapshot repository path")
+		}
+		settings["path"] = filepath.Join(path, suffix)
+	}
+	snapshotRepository["settings"] = settings
+
+	return snapshotRepository, nil
+}
```

Comment:
Could we please get some unit tests to lock in the behavior of this?

Also, what about the `default` case here? Should we return an error when the type of repository is invalid?

Reply:
> Could we please get some unit tests to lock in the behavior of this?

Peter is right, `mutateSnapshotRepositorySettings` is tested via `Test_updateState`. I can comment:

```
// Test_updateState tests updateState and consequently mutateSnapshotRepositorySettings.
```

> what about the default case here?
> Should we return an error when the type of repository is invalid?

I chose not to fully map all snapshot repository types and let Elasticsearch do the validation if an invalid type is used.